### ISSUE TITLE
6.0: TLS connection fails when SslProtocols.None is unavailable or effectively disabled

### ISSUE THREAD

**Issue by BasieP** [rabbitmq/rabbitmq-dotnet-client#764]
I tried upgrading to the rabbitmq.client version 6 (latest RC).

After i upgraded the client all my code was the same, but i got the BrokerUnreachable exception.

After some investigation i figured out that the ConnectionFactory no longer recognises uri's that start with amqps:// as amqps.

I feed my ConnectionFactory just a complete Uri.
In my case:  amqps://user:pass@myserver/myVhost

When i check the properties of the ConnectionFactory object i just created i see that he fills in most fields correctly based on the connectionstring i supplied, except for the protocol.

the 'Endpoint' property now is like this:
"amqp://myserver:5671"

so short version:
I can't use amqps with a connectionstring in the new client. And I could in the old one.

**Comment by michaelklishin** [comment]
I have added assertions on whether TLS is enabled on a `ConnectionFactory` in d8c3020beb4dd35dc109c9afeebc989fc9eb7fce and cannot reproduce. Please put together a small executable example.

**Comment by michaelklishin** [comment]
The suite above does not require a RabbitMQ node to be running locally and can be executed using

```sh
dotnet test projects/client/Unit --filter "FullyQualifiedName~RabbitMQ.Client.Unit.TestAmqpUri"
```

**Comment by michaelklishin** [comment]
@BasieP an effectively identical case is covered in `RabbitMQ.Client.Unit.TestAmqpUri`. Please provide evidence (e.g. a dump of object properties and server logs or a traffic capture) of your findings. It would also be great if you could run the above test in your environment.

**Comment by michaelklishin** [comment]
Also it would be incredibly helpful if this report provided a reasonably complete amount of information. Without it our team can only guess what is going on (at least given the above test suite modifications) and guessing is a very expensive way to spend our time.

Things such as

 * *Complete* exception stack trace
 * Server logs that would [contain connection event messages](https://www.rabbitmq.com/logging.html#connection-lifecycle-events)
 * An executable example e.g. in a public repository on GitHub

**Comment by BasieP** [comment]
Hi there,

I don't quite understand...
It's very easy to reproduce. It's just 3 lines i posted above:

I don't have access to server logs, but the server works. (i can connect with older clients)
The exact version i use of the rabbitmq client is: 6.0.0-pre9
The entire stacktrace:
```
RabbitMQ.Client.Exceptions.BrokerUnreachableException
  HResult=0x80131620
  Message=None of the specified endpoints were reachable
  Source=RabbitMQ.Client
  StackTrace:
   at RabbitMQ.Client.ConnectionFactory.CreateConnection(IEndpointResolver endpointResolver, String clientProvidedName)
   at Uniscope.Client.DebugConsole.Program.Main(String[] args) in D:\*********\Client.Console\Program.cs:line 25

  This exception was originally thrown at this call stack:
	System.Net.Security.SslState.ValidateCreateContext(bool, string, System.Security.Authentication.SslProtocols, System.Security.Cryptography.X509Certificates.X509Certificate, System.Security.Cryptography.X509Certificates.X509CertificateCollection, bool, bool, bool)
	System.Net.Security.SslStream.BeginAuthenticateAsClient(string, System.Security.Cryptography.X509Certificates.X509CertificateCollection, System.Security.Authentication.SslProtocols, bool, System.AsyncCallback, object)
	System.Net.Security.SslStream.AuthenticateAsClientAsync.AnonymousMethod__0(System.AsyncCallback, object)
	System.Net.Security.SslStream.AuthenticateAsClientAsync(string, System.Security.Cryptography.X509Certificates.X509CertificateCollection, System.Security.Authentication.SslProtocols, bool)
	RabbitMQ.Client.SslHelper.TcpUpgrade(System.IO.Stream, RabbitMQ.Client.SslOption)
	RabbitMQ.Client.Impl.SocketFrameHandler.SocketFrameHandler(RabbitMQ.Client.AmqpTcpEndpoint, System.Func<System.Net.Sockets.AddressFamily, RabbitMQ.Client.ITcpClient>, System.TimeSpan, System.TimeSpan, System.TimeSpan)
	RabbitMQ.Client.ConnectionFactory.CreateFrameHandler(RabbitMQ.Client.AmqpTcpEndpoint)
	RabbitMQ.Client.EndpointResolverExtensions.SelectOne<T>(RabbitMQ.Client.IEndpointResolver, System.Func<RabbitMQ.Client.AmqpTcpEndpoint, T>)

Inner Exception 1:
AggregateException: One or more errors occurred.

Inner Exception 2:
ArgumentException: The specified value is not valid in the 'SslProtocolType' enumeration.
Parameter name: sslProtocolType
```

**Comment by michaelklishin** [comment]
The exception originates in `System.Net.Security.SslState.ValidateCreateContext` and there's clear evidence that a TLS upgrade is performed so the hypothesis that `ConnectionFactory` TLS settings are not set correctly does not check out.

**Comment by BasieP** [comment]
> The exception originates in `System.Net.Security.SslState.ValidateCreateContext` and there's clear evidence that a TLS upgrade is performed so the hypothesis that `ConnectionFactory` TLS settings are not set correctly does not check out.

I don't really care where the error is accually.
Is there anything I can do to help you guys fix it?

**Comment by michaelklishin** [comment]
You don't have to care but either provide the relevant information we are asking for or at least refrain from making claims that the issue is in area X without providing any evidence of that. So for we have spent about an hour chasing an issue around `ConnectionFactory` state and URI parsing and when a complete stack trace was finally shared, that claim turned out to be false.

I will try to reproduce e.g. by modifying .NET client tutorials to connect to a TLS endpoint.

**Comment by BasieP** [comment]
> You don't have to care but either provide the relevant information we are asking for or at least refrain from making claims that the issue is in area X without providing any evidence of that. So for we have spent about an hour chasing an issue around `ConnectionFactory` state and URI parsing and when a complete stack trace was finally shared, that claim turned out to be false.
> 
> I will try to reproduce e.g. by modifying .NET client tutorials to connect to a TLS endpoint.

I'm sure you mean:
"Thank you for reporting a critical bug in our software, the problem was not where you think it was, but thanks anyway."

Again, if you want me to test something out, just let me know.

**Comment by michaelklishin** [comment]
We do appreciate people trying our preview builds and reporting their findings. We particularly appreciate reports with relevant information and some consideration for the time our team spends maintaining this software that you and other users get to use for free.

**Comment by lukebakken** [comment]
Here's the actual error I believe:

```
ArgumentException: The specified value is not valid in the 'SslProtocolType' enumeration.
Parameter name: sslProtocolType
```

I will investigate it and will modify the title / content of the issue if necessary.

@BasieP - what .NET environment are you using and on what operating system?

**Comment by BasieP** [comment]
I'm building our package for both netstandard2.0 and net462.
I tested it (with the above error) on a net462 application on Windows 10 (latest)

I'm trying to replicate on .net Core 3.1 as we speak

**Comment by lukebakken** [comment]
Thanks, I believe that is the same as our test environments (https://ci.appveyor.com/project/rabbitmq/rabbitmq-dotnet-client/history, https://travis-ci.org/github/rabbitmq/rabbitmq-dotnet-client). We also have an internal CI that runs on Debian Buster and uses these packages:

```
      dotnet-runtime-2.1 \
      dotnet-sdk-3.1 \
```

**Comment by michaelklishin** [comment]
Some relevant background: in https://github.com/rabbitmq/rabbitmq-dotnet-client/commit/fe39d6e5f94f4286d3763bc77f7a0871b7a8d756#diff-4777f22d87744f15542febe09ece4bbe we have changed the default TLS version used by connections from TLSv1.0 to`None` which means "[let the OS choose the version](https://docs.microsoft.com/en-us/dotnet/api/system.security.authentication.sslprotocols?view=netcore-2.2)".

https://github.com/robinrodricks/FluentFTP/issues/452 and https://github.com/robinrodricks/FluentFTP/issues/356 contain some interesting observations from another project that tried doing the same a few months earlier.

**Comment by michaelklishin** [comment]
Also according to https://github.com/robinrodricks/FluentFTP/issues/452#issuecomment-563965519 and https://referencesource.microsoft.com/#System/net/System/Net/SecureProtocols/_SslState.cs,164, it is possible to disable system default TLS versions. I could not find how we can verify if this is the case in this specific environment.

**Comment by BasieP** [comment]
Just for the record, I've not got a modified windows version. It's a private laptop, without corporate scripts that modify the configuration of it.
I run only windows defender, no other virus scanners and never modified my tls/ssl settings (as far as i'm aware)
I did load some root certificates in my store, but those are for https only and for different domains.

**Comment by michaelklishin** [comment]
Curiously https://github.com/mysql-net/MySqlConnector/commit/715c3b542a1862b308f8afbb535ba1fee34afb93 special cases handling of this exception to fall back to explicitly enabling TLSv1.1 and TLSv1.2 for the connection since `SslProtocols.None` is unsupported. While that targets an older .NET Core version, we perhaps should consider doing the same as the matrix of possible platforms and scenarios is still substantial, even with required .NET version bump in 6.0 [of this client].

**Comment by lukebakken** [comment]
Great find @michaelklishin. I will use that code here unless you beat me to it. I'm debugging an LDAP issue in Slack right now.

**Comment by michaelklishin** [comment]
It would be a pretty ugly hack in this client's architecture as we'd have to modify TLS options of an already instantiated `ConnectionFactory` after a connection attempt. I'm investigating.

**Comment by michaelklishin** [comment]
It's ironic that [Microsoft's own docs](https://docs.microsoft.com/en-us/dotnet/framework/network-programming/tls?view=netframework-4.6.2) on TLS best practices suggest exactly what we do but seemingly only for .NET 4.7+.

@lukebakken @bording @danielmarbach @stebet  should we consider bumping the minimum required version to .NET 4.7 instead of introducing ugly hacks  (https://github.com/rabbitmq/rabbitmq-dotnet-client/issues/764#issuecomment-602670966) to this client?

**Comment by michaelklishin** [comment]
@BasieP according to several issues I could find, several Microsoft guides and blog posts, what this client does in 6.0 should not only be safe but also the recommended and right thing to do starting with .NET 4.7. Can you try reproducing on .NET 4.7 and latest .NET Core? Also, can you try to explicitly set TLS versions enabled using something like this?

``` csharp
using System.Security.Authentication;

// ...

cf.Ssl.Versions = SslProtocols.Tls12;
```

(or similar)?

**Comment by stebet** [comment]
That'd certainly be ideal, if that's fine with you guys.

**Comment by bording** [comment]
The other thing to try if running on `net462` would be to set the [AppContext](https://docs.microsoft.com/en-us/dotnet/framework/network-programming/tls?view=netframework-4.6.2#for-net-framework-46---462-and-not-wcf) mentioned in the document. Seems like that is required to get the `None` behavior to work properly. Verifying that works would be good to know.

**Comment by bording** [comment]
> @lukebakken @bording @danielmarbach @stebet should we consider bumping the minimum required version to .NET 4.7 instead of introducing ugly hacks (#764 (comment)) to this client?

Overall I'd be okay with it, but actually doing that is rather tricky, because `netstandard2.0` supports `net461`, so even if you change the target frameworks to something like `net472;netstandard2.0`, it will still work on `net461` by choosing the `netstandard2.0` assembly, which doesn't solve anything.

The only real way to prevent that from happening is to add some MSbuild logic to the package that prevents it from compiling if  < `net472` is targeted. That doesn't stop people from installing the package, but it does prevent people from actually using it.

**Comment by michaelklishin** [comment]
We can explicitly refuse e.g. to open a connection if an incompatible version is detected. We do that in `rabbitmqadmin` for old (and now entirely unmaintained) Python 2 versions because they don't provide certain TLS API elements. That would be drastic but with TLS, if something is broken/incomplete/unavailable in a certain version, the solutions are either to require a newer version or employ ugly workarounds.

Let me finish the PR so we can decide if it'd be considered appropriate by other committers.

**Comment by BasieP** [comment]
> ```cs
> using System.Security.Authentication;
> 
> // ...
> 
> cf.Ssl.Version = SslProtocols.Tls12;
> ```

yeah that works. But i have the same issues with .Net core 3.1 (latest greatest)
So when bumbing minimal requirements for .net framework, i'm wondering what this implies for .net core 3.1.
Any thoughts on that?

**Comment by michaelklishin** [comment]
If .NET Core 3.1 is affected we'd have to find a solution other than simply requiring a newer version.

**Comment by michaelklishin** [comment]
I am by no means a .NET expert but there's this curious recommendation for .NET 4.6 in [this security best practices document](https://docs.microsoft.com/en-us/dotnet/framework/network-programming/tls?view=netframework-4.6.2):

> Set the `DontEnableSystemDefaultTlsVersions` AppContext switch to `false`

and the [document it refers to](https://docs.microsoft.com/en-us/dotnet/framework/network-programming/tls?view=netframework-4.6.2#configuring-security-via-appcontext-switches) says:

> A value of `false` for `Switch.System.Net.DontEnableSystemDefaultTlsVersions` causes your 
> app to allow the operating system to choose the protocol. A value of `true` causes your app to 
> use protocols picked by the .NET Framework.

So this behavior is also app context-controlled, which is useful to know as we must mention this in our TLS guide.

**Comment by michaelklishin** [comment]
So there are several different "scenarios" that we should consider:

 * .NET 4.6.1
 * .NET 4.7
 * .NET Core 3.x
 * `Switch.System.Net.DontEnableSystemDefaultTlsVersions` set to `false`
 * `Switch.System.Net.DontEnableSystemDefaultTlsVersions` set to `true`

So even with recent, almost cutting edge .NET framework versions, `SslProtocols.None` can behave differently from what we expect/be effectively disabled. Is my understanding correct?

**Comment by lukebakken** [comment]
So much 🤦‍♂ . I thought I was doing the correct thing when I switched the default to `None` 🙄

**Comment by michaelklishin** [comment]
@BasieP can you please give https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/766 a try with .NET 4.6.1 as well as .NET Core 3?

**Comment by BasieP** [comment]
How do i do that? I am using nuget now,

Is there new beta version? Or should i compile the pr? I could give that a
try in a couple of hours..

**Comment by michaelklishin** [comment]
There are no NuGet releases for PRs. @lukebakken @bording @stebet what's the most common approach for depending on a library built locally from source?

**Comment by danielmarbach** [comment]
If the nuget is a build artifact it can be copied into a local folder on disk that can be used as a package sources.

**Comment by BasieP** [comment]
The problem is the compiling. I have the source on disk now, but it doesn't compile in visual studio..
```
Source file 'D:\workspace\rabbitmq-dotnet-client\projects\client\RabbitMQ.Client\..\..\..\gensrc\autogenerated-api-0-9-1.cs' could not be found.
```

Prolly there are lots of steps you do on your buildserver that are hard for me to reproduce.
Best practices are to do as less as posible on buildservers.. Msbuild can do lots in itself..

**Comment by michaelklishin** [comment]
There aren't many steps: you must restore the project and then there's one code generation phase. See [Running Tests](https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/master/RUNNING_TESTS.md) and `build.bat` in the repo root. Once restore and codegen succeed,

```
dotnet build
```

is the only thing you need to run during development.

**Comment by michaelklishin** [comment]
The only thing Running Tests above does not mention is that codegen requires Python to be installed. I'm attaching compiled artifacts of the actual client (so, no codegen or tests) produced by `dotnet build`. Note that I do not use Windows but in theory it should not matter much these days?

[dotnet-client-debug-304f97684.zip](https://github.com/rabbitmq/rabbitmq-dotnet-client/files/4371435/dotnet-client-debug-304f97684.zip)

**Comment by michaelklishin** [comment]
and a package produced on the #766 branch by `dotnet pack`:

[RabbitMQ.Client.nupkg-304f97684.zip](https://github.com/rabbitmq/rabbitmq-dotnet-client/files/4371444/RabbitMQ.Client.nupkg-304f97684.zip)

**Comment by michaelklishin** [comment]
We have added some initial docs in https://github.com/rabbitmq/rabbitmq-website/commit/6376dbb19a3ef82bdb6918a3b04a58f34f738ec1. They already mention 6.0 by design:

 * We hope to ship it in the next few weeks
 * We'd like to stress that default TLS version pick will different between client, .NET and OS versions, as it can be a deal-breaker for successful connectivity, and not only in the way we see reported here.

They will go live shortly.

**Comment by lukebakken** [comment]
> somehow the BasicDeliverEventArgs.Body is now of type System.ReadOnlyMemory

@BasieP version 6.0 has a dependency on the following package:

https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/master/projects/client/RabbitMQ.Client/RabbitMQ.Client.csproj#L46

```
<PackageReference Include="System.Memory" Version="4.5.3" />
```

https://www.nuget.org/packages/System.Memory/

Version 6.0.0 has dramatic memory allocation improvements that this package makes possible.

**Comment by lukebakken** [comment]
PR builds are pushed to myget - https://www.myget.org/feed/rabbitmq/package/nuget/RabbitMQ.Client

`RabbitMQ.Client.6.0.0-pr-56b16ba6.nupkg` is the latest build of this PR. I'll see if I can improve the package name.

**Comment by michaelklishin** [comment]
#766 is in and there will be a new preview build of 6.0 shortly. We assume we did everything we could in the meantime and the docs were updated to mention what can affect protocol version selection. Explicitly specifying a protocol version in environments where automatic selection is disabled is all that has to be done.

**Comment by BasieP** [comment]
i've never been helped faster on github 👍

**Comment by michaelklishin** [comment]
@BasieP thank you for reporting a really intricate issue. While I can't say how effective our workaround is going to be at the moment (given the number of variables involved), we now understand the problem with TLS version selection and documentation updates alone would save a lot of people some time and frustration.

**Comment by michaelklishin** [comment]
[6.0.0-pre.14](https://www.nuget.org/packages/RabbitMQ.Client/6.0.0-pre.14) should include #766 and more. @BasieP please give it a try as we are very close to cutting an RC.

**Comment by BasieP** [comment]
like a charm :)
thanks guys
